### PR TITLE
Nft pool updates

### PR DESCRIPTION
**Pull request by grandizzy**
- slither fixes
- any actor may recollateralize or repay on behalf of another borrower
- test updates

### CODE REVIEW

File: src/erc721/ERC721Pool.sol [hunk 1]
```diff
@@ -172,17 +163,14 @@ contract ERC721Pool is IERC721Pool, ScaledPool {
         _updateInterestRate(curDebt, curLup);
 
         // remove tokenIds and transfer to caller
-        for (uint i; i < tokenIds_.length;) {
+        for (uint256 i = 0; i < tokenIds_.length;) {
```
Comment (MikeHathaway) [inline src/erc721/ERC721Pool.sol:167]: What's the purpose of initializing the variable to 0? Isn't that handled by default? Is that due to a slither warning?

Comment (grandizzy) [reply src/erc721/ERC721Pool.sol:167]: Yeah, slither related. No gas difference though

File: src/erc721/ERC721Pool.sol [hunk 2]
```diff
@@ -245,28 +233,25 @@ contract ERC721Pool is IERC721Pool, ScaledPool {
...
         // move required collateral from sender to pool
-        for (uint i; i < tokenIds_.length;) {
-            if (_tokenIdsAllowed.length() != 0) {
-                require(_tokenIdsAllowed.contains(tokenIds_[i]), "P:ONLY_SUBSET");
-            }
-
-            _bucketCollateralTokenIds.add(tokenIds_[i]);
+        for (uint256 i = 0; i < tokenIds_.length;) {
+            if (_tokenIdsAllowed.length() != 0)                     require(_tokenIdsAllowed.contains(tokenIds_[i]),     "P:ONLY_SUBSET");
+            if (!_bucketCollateralTokenIds.contains(tokenIds_[i])) require(_bucketCollateralTokenIds.add(tokenIds_[i]), "P:ADD_BC_FAIL");
```
Comment (MikeHathaway) [inline src/erc721/ERC721Pool.sol]: I still don't fully understand why we need to wrap `.add()` in a `contains()` check... `add()` calls `_contains()` internally. This seems like a lot of unnecessary gas expenditure

Comment (grandizzy) [reply src/erc721/ERC721Pool.sol]: we can remove contains check but I still think require should stay and be sure tokens are added. Contains should prevent failures when token is already added.

Comment (MikeHathaway) [reply src/erc721/ERC721Pool.sol]: I'm okay with leaving the require around the `_add()` operation, but definitely think the outer `contains()` check is unnecessary

Comment (grandizzy) [reply src/erc721/ERC721Pool.sol]: Makes sense, will remove

Comment (grandizzy) [reply src/erc721/ERC721Pool.sol]: removed with https://github.com/ajna-finance/contracts/pull/217/commits/277943f21704d601db2935e8a2317b5b17ef975b

**Review by MikeHathaway — APPROVED** [review]
LGTM! Thanks for the changes

**Pull request merged**